nncf/experimental/torch/nas/bootstrapNAS/training/cosine_lr_scheduler.py

```diff
@@ -0,0 +1,116 @@
+
+
+def adjust_learning_rate(optimizer, epoch, init_lr, epochs, batch=0, nBatch=None, lr_schedule_type='cosine'):
+    """ adjust learning of a given optimizer and return the new learning rate """
+    new_lr = calc_learning_rate(epoch, init_lr, epochs, batch, nBatch, lr_schedule_type)
+    for param_group in optimizer.param_groups:
+        param_group['lr'] = new_lr
+    return new_lr
+
+
+def warmup_adjust_learning_rate(optimizer, T_total, nBatch, epoch, batch=0, warmup_lr=0):
+    T_cur = epoch * nBatch + batch + 1
+    new_lr = T_cur / T_total * (self.init_lr - warmup_lr) + warmup_lr
+    for param_group in optimizer.param_groups:
+        param_group['lr'] = new_lr
+    return new_lr
+
+
+def calc_learning_rate(epoch, init_lr, n_epochs, batch=0, nBatch=None, lr_schedule_type='cosine'):
+    if lr_schedule_type == 'cosine':
+        t_total = n_epochs * nBatch
+        t_cur = epoch * nBatch + batch
+        lr = 0.5 * init_lr * (1 + math.cos(math.pi * t_cur / t_total))
+    elif lr_schedule_type is None:
+        lr = init_lr
+    else:
+        raise ValueError('do not support: %s' % lr_schedule_type)
+    return lr
+
+
+class CosineLRScheduler(BaseCompressionScheduler):
+    def __init__(self, optimizer, num_steps_in_epoch, *, base_lr, num_epochs, warmup_epochs=0, warmup_lr=None):
+        super().__init__()
+        self._base_lr = base_lr
+        self._num_epochs = num_epochs
+        self._warmup_epochs = warmup_epochs
+        self._warmup_lr = warmup_lr
+        self._optimizer = optimizer
+        self._num_steps_in_epoch = num_steps_in_epoch
+
+    def step(self, next_step: Optional[int] = None) -> None:
+        super().step(next_step)
+        step_from_epoch_start = self.current_step - (self.current_epoch*(self._num_steps_in_epoch+1))
+        if self.current_epoch < self._warmup_epochs:
+            new_lr = warmup_adjust_learning_rate(optimizer=self._optimizer,
+                                                 T_total=self._warmup_epochs * self._num_steps_in_epoch,
+                                                 nBatch=self._num_steps_in_epoch,
+                                                 epoch=self.current_epoch,
+                                                 batch=step_from_epoch_start,
+                                                 warmup_lr=self._warmup_lr)
+        else:
+            new_lr = adjust_learning_rate(optimizer=self._optimizer,
+                                          epoch=self.current_epoch - self._warmup_epochs,
+                                          init_lr=self._base_lr,
+                                          epochs=self._num_epochs,
+                                          batch=step_from_epoch_start,
+                                          nBatch=self._num_steps_in_epoch,
+                                          lr_schedule_type='cosine')
+
+    @BaseCompressionScheduler.current_epoch.setter
+    def current_epoch(self, val: float) -> NoReturn:
```

Comment:
It's not correct to override the signature of the method. It returns int in the base class.

Reply:
We can replace float to int in the type hinting. However, according to PEP 0484 https://peps.python.org/pep-0484/#the-numeric-tower it seems that we should use float as type hint. Thanks!

Reply:
I'm not asking to replace float with int. I'm telling about return type. This method returns int in the base class, `NoReturn` changes the signature, which is not correct